File: backend/app/models/ride.py
from sqlalchemy import Float, ForeignKey, String
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.database import Base


class Ride(Base):
    __tablename__ = "rides"

    id: Mapped[int] = mapped_column(
        primary_key=True,
        index=True,
    )

    customer_id: Mapped[int] = mapped_column(
        ForeignKey("users.id"),
        nullable=False,
    )

    driver_id: Mapped[int | None] = mapped_column(
        ForeignKey("drivers.id"),
        nullable=True,
    )

    pickup_lat: Mapped[float] = mapped_column(
        Float,
        nullable=False,
    )

    pickup_lng: Mapped[float] = mapped_column(
        Float,
        nullable=False,
    )

    destination_lat: Mapped[float] = mapped_column(
        Float,
        nullable=False,
    )

    destination_lng: Mapped[float] = mapped_column(
        Float,
        nullable=False,
    )

    fare: Mapped[float] = mapped_column(
        Float,
        nullable=False,
    )

    status: Mapped[str] = mapped_column(
        String(30),
        default="searching",
        nullable=False,
    )

    customer = relationship(
        "User",
        back_populates="rides",
    )

    driver = relationship(
        "Driver",
        back_populates="rides",
    )

    payment = relationship(
        "Payment",
        back_populates="ride",
        uselist=False,
    )

    rating = relationship(
        "Rating",
        back_populates="ride",
        uselist=False,
    )

File: customer_app/lib/models/ride.dart
class Ride {
  final int id;
  final int customerId;
  final int? driverId;

  final double pickupLat;
  final double pickupLng;

  final double destinationLat;
  final double destinationLng;

  final double fare;
  final String status;

  Ride({
    required this.id,
    required this.customerId,
    required this.driverId,
    required this.pickupLat,
    required this.pickupLng,
    required this.destinationLat,
    required this.destinationLng,
    required this.fare,
    required this.status,
  });

  factory Ride.fromJson(Map<String, dynamic> json) {
    return Ride(
      id: json["id"],
      customerId: json["customer_id"],
      driverId: json["driver_id"],
      pickupLat: (json["pickup_lat"] as num).toDouble(),
      pickupLng: (json["pickup_lng"] as num).toDouble(),
      destinationLat:
          (json["destination_lat"] as num).toDouble(),
      destinationLng:
          (json["destination_lng"] as num).toDouble(),
      fare: (json["fare"] as num).toDouble(),
      status: json["status"],
    );
  }
}